File: LeetCode/cpp/138.cpp
class Solution {
public:
    RandomListNode *copyRandomList(RandomListNode *head) {
        if (!head)
            return nullptr;
        auto node = head;
        while (node) {
            auto next = new RandomListNode(node->label);
            next->next = node->next;
            node->next = next;
            node = node->next->next;
        }
        node = head;
        auto next = node->next, ret = next;
        while (node) {
            if (node->random)
                next->random = node->random->next;
            node = node->next->next;
            next = node ? node->next : nullptr;
        }
        node = head;
        while (node) {
            auto temp = node->next;
            node->next = node->next->next;
            temp->next = node->next ? node->next->next : nullptr;
            node = node->next;
        }
        return ret;
    }
};

File: LeetCode/python/138.py
class Solution(object):
    def copyRandomList(self, head):
        """
        :type head: RandomListNode
        :rtype: RandomListNode
        """
        if not head:
            return head
        newHead = RandomListNode(0)
        currentNode, curr = newHead, head
        nodeDict = {}
        while curr:
            tempNode = RandomListNode(curr.label)
            nodeDict[curr] = tempNode
            currentNode.next = tempNode
            currentNode, curr = currentNode.next, curr.next
        curr, currentNode = head, newHead.next
        while curr and currentNode:
            if curr.random:
                currentNode.random = nodeDict[curr.random]
            curr, currentNode = curr.next, currentNode.next
        return newHead.next
